url4fdf79ae2c_beckchr_juel_tgz-pJ8-de_odysseus_el_samples_xml_sax_XMELFilterJ8/src/de/odysseus/el/tree/impl/ast/AstFunction.java

```diff
@@ -0,0 +1,166 @@
+
+public class AstFunction extends AstRightValue implements FunctionNode {
+	private final int index;
+	private final String name;
+	private final AstParameters params;
+	private final boolean varargs;
+
+	public AstFunction(String name, int index, AstParameters params) {
+		this(name, index, params, false);
+	}
+
+	public AstFunction(String name, int index, AstParameters params, boolean varargs) {
+		this.name = name;
+		this.index = index;
+		this.params = params;
+		this.varargs = varargs;
+	}
+
+	/**
+	 * Invoke method.
+	 * @param bindings
+	 * @param context
+	 * @param base
+	 * @param method
+	 * @return method result
+	 * @throws InvocationTargetException
+	 * @throws IllegalAccessException
+	 */
+	protected Object invoke(Bindings bindings, ELContext context, Object base, Method method)
+		throws InvocationTargetException, IllegalAccessException {
+		Class<?>[] types = method.getParameterTypes();
+		Object[] params = null;
+		if (types.length > 0) {
+			params = new Object[types.length];
+			if (varargs && method.isVarArgs()) {
+				for (int i = 0; i < params.length - 1; i++) {
+					Object param = getParam(i).eval(bindings, context);
+					if (param != null || types[i].isPrimitive()) {
+						params[i] = bindings.convert(param, types[i]);
+					}
+				}
+				int varargIndex = types.length - 1;
+				Class<?> varargType = types[varargIndex].getComponentType();
+				int length = getParamCount() - varargIndex;
+				Object array = null;
+				if (length == 1) { // special: eventually use argument as is
+					Object param = getParam(varargIndex).eval(bindings, context);
+					if (param != null && param.getClass().isArray()) {
+						if (types[varargIndex].isInstance(param)) {
+							array = param;
+						} else { // coerce array elements
+							length = Array.getLength(param);
+							array = Array.newInstance(varargType, length);
+							for (int i = 0; i < length; i++) {
+								Object elem = Array.get(param, i);
+								if (elem != null || varargType.isPrimitive()) {
+									Array.set(array, i, bindings.convert(elem, varargType));
+								}
+							}
+						}
+					} else { // single element array
+						array = Array.newInstance(varargType, 1);
+						if (param != null || varargType.isPrimitive()) {
+							Array.set(array, 0, bindings.convert(param, varargType));
+						}
+					}
+				} else {
+					array = Array.newInstance(varargType, length);
+					for (int i = 0; i < length; i++) {
+						Object param = getParam(varargIndex + i).eval(bindings, context);
+						if (param != null || varargType.isPrimitive()) {
+							Array.set(array, i, bindings.convert(param, varargType));
+						}
+					}
+				}
+				params[varargIndex] = array;
+			} else {
+				for (int i = 0; i < params.length; i++) {
+					Object param = getParam(i).eval(bindings, context);
+					if (param != null || types[i].isPrimitive()) {
+						params[i] = bindings.convert(param, types[i]);
+					}
+				}
+			}
+		}
+		return method.invoke(base, params);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Catch `InvocationTargetException` explicitly when you call `Method.invoke()`. Use `Throwable.getCause()` or `getTargetException()` in the catch handler to access more information about the underlying exception. See the following diffs on GitHub that make this code change: [diff1](https://github.com/EsfingeFramework/gamification/commit/20792047e6c69457cca03e877f2ada578188f32a#diff-6808c12fe9ddaea042db84f95dd5ba78adab99d0cb18bb4d85916a8723530768L21), [diff2](https://github.com/cucumber-attic/cuke4duke/commit/f4821d400cf2590126f1298a8dff9f7a33d45dc5#diff-80adaffc20c0e6e7d2e42312386ffe1ddc10213bf5b8e8919895fd85d666c8deL34)

Learn more: [InvocationTargetException in the Java Platform, Standard Edition 8 API Specification](https://docs.oracle.com/javase/8/docs/api/java/lang/reflect/InvocationTargetException.html)